pkg/hostpath/hostpath.go

```diff
@@ -818,3 +830,57 @@ func (hp *hostPath) loadVolsFromDisk() error {
 		return nil
 	})
 }
+
+// Prune inspects all the volumes stored on disk and sees if their associated pods still exists.  If not, the volume
+// file in question is deleted from disk.
+func (hp *hostPath) Prune(kubeClient kubernetes.Interface) {
+	filesToPrune := []string{}
+	filepath.Walk(hp.volMapRoot, func(path string, info os.FileInfo, err error) error {
+		if info == nil {
+			return nil
+		}
+		if err != nil {
+			// continue to next file
+			return nil
+		}
+		if info.IsDir() {
+			return nil
+		}
+		fileName := filepath.Join(hp.volMapRoot, info.Name())
+		dataFile, oerr := os.Open(fileName)
+		if oerr != nil {
+			klog.V(0).Infof("loadVolsFromDisk error opening file %s: %s", fileName, err.Error())
+			// continue to next file
+			return nil
+		}
+		dataDecoder := json.NewDecoder(dataFile)
+		hpv := &hostPathVolume{}
+		err = dataDecoder.Decode(hpv)
+		if err != nil {
+			klog.V(0).Infof("loadVolsFromDisk error decoding file %s: %s", fileName, err.Error())
+			// continue to next file
+			return nil
+		}
+		if hpv == nil {
+			klog.V(0).Infof("loadVolsFromDisk nil but no error for file %s", fileName)
+			// continue to next file
+			return nil
+		}
+		hpv.Lock = &sync.Mutex{}
+		_, err = kubeClient.CoreV1().Pods(hpv.GetPodNamespace()).Get(context.TODO(), hpv.GetPodName(), metav1.GetOptions{})
+		if err != nil && kerrors.IsNotFound(err) {
+			klog.V(2).Infof("pruner: hpv %q: %s", fileName, err.Error())
+			filesToPrune = append(filesToPrune, fileName)
```

Comment:
Why not just attempt to delete the file(s) right after you check if the pod exists or not, instead of iterating over a separate slice when you are done?

Reply:
I was paranoid about messing up the walk function if we deleted items from the list it was iterating over while we were in the middle of iterating (old Java pain points ;-)) 

Trying some internet searches now to confirm or deny that paranoia ....

Reply:
found nothing conclusive - given this runs only once per 10 minutes, I'm leaning toward caution and siding toward paranoia vs. efficiency

I'm putting a comment to explain rationale

that work for you?

Reply:
Yep, that works for me.

Reply:
cool thanks